fix issue with different number of elements in snmp table fields

michep opened the pull request:
fix issue with different number of elements in snmp table fields:

snmpwalk -v 2c -c AABOZy8TUJkD 10.0.0.4 ifName |wc -l
52
snmpwalk -v 2c -c AABOZy8TUJkD 10.0.0.4 ifInMulticastPkts |wc -l
48

michep committed:
fix issue with different number of elements in snmp table fields:

snmpwalk -v 2c -c AABOZy8TUJkD 10.0.0.4 ifName |wc -l
52
snmpwalk -v 2c -c AABOZy8TUJkD 10.0.0.4 ifInMulticastPkts |wc -l
48

PatrykMatyjasek commented:
Thanks for your contribution
@katarzyna-z please take a look :)

katarzyna-z commented on collector/collector.go:
```diff
@@ -392,17 +407,11 @@ func getDynamicNamespaceElements(handler *snmpgo.SNMP, results []*snmpgo.VarBind
 				}
-				metric.Namespace[i].Values = append(metric.Namespace[i].Values, oidParts[metric.Namespace[i].OidPart])
+				idx := oidParts[metric.Namespace[i].OidPart]
+				iidx, _ := strconv.Atoi(idx) //TODO: error handling?
```
Please add the error handling ;)

katarzyna-z commented on collector/collector.go:
```diff
@@ -512,3 +521,27 @@ func modifyNumericMetric(data interface{}, scale float64, shift float64) interfa
+			} else {
+				index = strings.Split(roidTrimmed, ".")[oidLen-1]
+			}
+			idx, _ := strconv.Atoi(index) //TODO: error handling?
```
Please add the error handling.

katarzyna-z commented on collector/collector_test.go:
```diff
@@ -763,7 +758,7 @@ var (
 			{"source": "snmp", "name": "snmp part", "description": "part received through SNMP request", "OID": ".1.3.6.1.2.1.1.9.1.3"},
 			{"source": "string", "string": "value"}
 		  ],
-		  "OID": ".1.3.6.1.2.1.1.9.1.4",
```
Why have you changed it? Is there any specific reason?

katarzyna-z replied:
@michep I've seen your comment but are you able to identify change in the code which triggers this change in tests?

michep commented:
I did that because in test SNMP results:
https://github.com/michep/snap-plugin-collector-snmp/blob/7896e1c3d96bd715cfcbfc70502a251bc6ee2a1c/collector/collector_test.go#L87
there's only values for OID ".1.3.6.1.2.1.1.9.1.3.1".
And if in metric configuration, in metric value OID or dynаmic namespace
OID, we reference to undefined OID, like ".1.3.6.1.2.1.1.9.1.4" then now
there will be an error in some test, now I dont remember in which.

On 6 July 2017 at 15:03, Katarzyna Kujawa <notifications@github.com> wrote:

> *@katarzyna-z* commented on this pull request.
> ------------------------------
>
> In collector/collector_test.go
> <https://github.com/intelsdi-x/snap-plugin-collector-snmp/pull/27#discussion_r125879765>
> :
>
> > @@ -763,7 +758,7 @@ var (
>  			{"source": "snmp", "name": "snmp part", "description": "part received through SNMP request", "OID": ".1.3.6.1.2.1.1.9.1.3"},
>  			{"source": "string", "string": "value"}
>  		  ],
> -		  "OID": ".1.3.6.1.2.1.1.9.1.4",
>
> Why have you changed it? Is there any specific reason?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/intelsdi-x/snap-plugin-collector-snmp/pull/27#pullrequestreview-48311177>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJmCYG5_JeB4lX_GLM7OFu7cxmWuV19Oks5sLM0egaJpZM4OMkHB>
> .
>



-- 
Михаил Чепайкин
skype: mchepaykin

katarzyna-z commented on collector/collector.go:
```diff
@@ -512,3 +521,27 @@ func modifyNumericMetric(data interface{}, scale float64, shift float64) interfa
+	for _, r := range results {
+		roidTrimmed := strings.Trim(r.Oid.String(), ".")
+		roidLen := len(strings.Split(roidTrimmed, "."))
+		if ((roidLen == oidLen) || (roidLen == oidLen+1)) && strings.HasPrefix(roidTrimmed, oidTrimmed) {
```
Could you explain this condition?

katarzyna-z replied:
@michep Thanks for detailed explanation of this condition.

Please notice that for both queries you have the same OID in response
```
# snmpwalk -v 2c -On -c public 10.0.0.4 .1.3.6.1.2.1.1.5
.1.3.6.1.2.1.1.5.0 = STRING: mfms_switch1

# snmpwalk -v 2c -On -c public 10.0.0.4 .1.3.6.1.2.1.1.5.0
.1.3.6.1.2.1.1.5.0 = STRING: mfms_switch1```

r.Oid.String() is the OID from device response so it should have the same form and it doesn't depend on OID in SNMP request.

Why in this function `oid string` is needed as an argument of `snmpResults2Map`? Taking into account that device replies with OID in full form (for request with .1.3.6.1.2.1.1.5 and .1.3.6.1.2.1.1.5.0 in response there is .1.3.6.1.2.1.1.5.0 as OID), you can split `roidTrimmed` and you use the last element as `index`, so the if condition is not needed.

It is what I see but please correct me if I'm wrong and you catch response from device in different form.

michep commented:
Sure. Func snmpResults2Map make a map of results from array. The key for
map will be snmp value index.

Lets consider this:

# snmpwalk -v 2c -c public 10.0.0.4 ifDescr| wc -l
52

# snmpwalk -v 2c -c public 10.0.0.4 ifDescr
IF-MIB::ifDescr.1 = STRING: Vlan1
IF-MIB::ifDescr.108 = STRING: Vlan108
IF-MIB::ifDescr.10101 = STRING: GigabitEthernet0/1
IF-MIB::ifDescr.10102 = STRING: GigabitEthernet0/2
<...>
IF-MIB::ifDescr.10148 = STRING: GigabitEthernet0/48
IF-MIB::ifDescr.10501 = STRING: Null0

# snmptranslate -On -IR ifDescr
.1.3.6.1.2.1.2.2.1.2

# snmpwalk -v 2c -c public 10.0.0.4 ifInMulticastPkts |wc -l
48

# snmpwalk -v 2c -c public 10.0.0.4 ifInMulticastPkts
IF-MIB::ifInMulticastPkts.10101 = Counter32: 17
IF-MIB::ifInMulticastPkts.10102 = Counter32: 0
<...>
IF-MIB::ifInMulticastPkts.10147 = Counter32: 3025000434
IF-MIB::ifInMulticastPkts.10148 = Counter32: 1895904384

# snmptranslate -On -IR ifInMulticastPkts
.1.3.6.1.2.1.31.1.1.1.2

and finally this part of setfile:

    {
        "mode": "table",
        "namespace": [
            {"source": "string", "string": "cisco"},
            {"source": "string", "string": "if-table"},
            {"source": "snmp", "name": "interface", "description":
"interface name", "OID": ".1.3.6.1.2.1.2.2.1.2"},
            {"source": "index", "name": "id", "description": "interface
index from OID", "oid_part": 11},
            {"source": "string", "string": "in_multicast"}
        ],
        "OID": ".1.3.6.1.2.1.31.1.1.1.2",
        "scale": 1.0,
        "shift": 0,
        "unit": "packets",
        "description": "number of inbound multicast packets"
    }

So, in this case we are looking for value of ifInMulticastPkts (OID =
.1.3.6.1.2.1.31.1.1.1.2) and want to use ifDescr (OID =
.1.3.6.1.2.1.2.2.1.2) as part of metrics namespace. But you see that the
number of ifInMulticastPkts field values less then number of ifDescr. This
means that we need tohave exact link, whichifDescr is related to which
ifInMulticastPkts.And it could be done by using "index" value in results
OIDs, like IF-MIB::ifDescr.10101 is corresponds to
IF-MIB::ifInMulticastPkts.10101 (10101 is that "index").

To create map with indexes as keys we need to process snmp results with
respect of the OID that we were actually looking for.

Take a look at this:

# snmptranslate -On -IR sysName
.1.3.6.1.2.1.1.5

# snmpwalk -v 2c -On -c public 10.0.0.4 .1.3.6.1.2.1.1.5
.1.3.6.1.2.1.1.5.0 = STRING: mfms_switch1

# snmpwalk -v 2c -On -c public 10.0.0.4 .1.3.6.1.2.1.1.5.0
.1.3.6.1.2.1.1.5.0 = STRING: mfms_switch1

This wes to show that we can query some snmp value by only its OID or by
OID plus index - result will be the same.

So, in my condition I check, that the resulting OID is has the same
starting part (strings.HasPrefix(roidTrimmed, oidTrimmed) as OID in query,
and the same length (roidLen == oidLen) or just one element larger (roidLen
== oidLen+1) than the OID in query.

I'm really hope that I've explained it clearly :)

On 6 July 2017 at 17:13, Katarzyna Kujawa <notifications@github.com> wrote:

> *@katarzyna-z* commented on this pull request.
> ------------------------------
>
> In collector/collector.go
> <https://github.com/intelsdi-x/snap-plugin-collector-snmp/pull/27#discussion_r125911376>
> :
>
> > @@ -512,3 +521,27 @@ func modifyNumericMetric(data interface{}, scale float64, shift float64) interfa
>  	}
>  	return modifiedData
>  }
> +
> +func snmpResults2Map(results []*snmpgo.VarBind, oid string) (map[int]*snmpgo.VarBind, serror.SnapError) {
> +	index := ""
> +	res := make(map[int]*snmpgo.VarBind)
> +	oidTrimmed := strings.Trim(oid, ".")
> +	oidLen := len(strings.Split(oidTrimmed, "."))
> +
> +	for _, r := range results {
> +		roidTrimmed := strings.Trim(r.Oid.String(), ".")
> +		roidLen := len(strings.Split(roidTrimmed, "."))
> +		if ((roidLen == oidLen) || (roidLen == oidLen+1)) && strings.HasPrefix(roidTrimmed, oidTrimmed) {
>
> Could you explain this condition?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/intelsdi-x/snap-plugin-collector-snmp/pull/27#pullrequestreview-48346427>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJmCYBNxrdNpVkNbWq_Pl4IAbx4CZLEJks5sLOuegaJpZM4OMkHB>
> .
>



-- 
Михаил Чепайкин
skype: mchepaykin

katarzyna-z commented:
Please add comment in this for to this condition: "we can query some snmp value by only its OID or by
OID plus index, snmpwalk query using .1.3.6.1.2.1.1.5 and .1.3.6.1.2.1.1.5.1 gives the same result".

katarzyna-z commented:
I'm continuing review of this code. I'm checking if this fix can be made easier or in different way. If I have any suggestions I'll share them.

katarzyna-z commented on collector/collector_test.go:
```diff
@@ -576,7 +569,8 @@ func TestGetDynamicNamespaceElements(t *testing.T) {
 					configReader.Namespace{Source: "string", String: "test2"},
-					configReader.Namespace{Source: "snmp", Oid: ".1.3.6.1.2.1.1.9.1.3", Name: "name"},
+					configReader.Namespace{Source: "snmp", Oid: ".1.3.6.1.2.1.1.9.1.3", Name: "name1"},
+					configReader.Namespace{Source: "snmp", Oid: ".1.3.6.1.2.1.1.9.1.7", Name: "name2"},
```
This test case is to test `getDynamicNamespaceElements` ```with incorrect configuration of dynamic elements of namespace, the number of results is not equal the number of dynamic elements of namespace``` in this pull request you are removing this condition, so this test case should be removed/replaced with different one.

katarzyna-z commented on collector/collector_test.go:
```diff
@@ -314,10 +311,8 @@ func TestCollectMetrics(t *testing.T) {
-				So(dynamics, ShouldContain, 3)
-				So(dynamics, ShouldContain, 4)
-				So(dynamics, ShouldContain, 5)
+				So(dynamics, ShouldHaveLength, 1)
```
Could you identify changes in the code which triggered this change in tests?

katarzyna-z commented:
You should consider this case:

User wants to collect metrics from this node:
```
$ snmpwalk -v 2c -c public 10.0.0.4 ifInMulticastPkts

IF-MIB::ifInMulticastPkts.10101 = Counter32: 17
IF-MIB::ifInMulticastPkts.10102 = Counter32: 0
IF-MIB::ifInMulticastPkts.10147 = Counter32: 3025000434
IF-MIB::ifInMulticastPkts.10148 = Counter32: 1895904384
```
but in configuration of namespace user has, as the namespace element, this node:
```
snmpwalk -v 2c -c public 10.0.0.4 ifDescr

IF-MIB::ifDescr.1 = STRING: GigabitEthernet0/1
IF-MIB::ifDescr.2 = STRING: GigabitEthernet0/2
IF-MIB::ifDescr.3 = STRING: GigabitEthernet0/2
```

so there are different indexes which don't match. 

Ideally you should add test to cover that case. 
If you considered this case and I don't see, please show me the error message which will be visible for this case.

katarzyna-z commented on collector/collector.go:
```diff
@@ -375,13 +382,21 @@ func getDynamicNamespaceElements(handler *snmpgo.SNMP, results []*snmpgo.VarBind
+			}
+
+			for idx, part := range partsMap {
+				if _, ok := resultMap[idx]; ok {
```
What if there is an element in `resultMap` for which there is not matching element in `partsMap`?

katarzyna-z commented:
You need to add clear description with examples about new method for creating metric namespace, probably [namespace](https://github.com/intelsdi-x/snap-plugin-collector-snmp#namespace) section will be good for that.

katarzyna-z commented:
If you need to modify code to address review comments, please do it in separate commits. This helps to review changes. You will need to squash commits into single one after approval of changes.
